realm/realm-library/src/main/java/io/realm/BaseRealm.java

```diff
@@ -282,6 +284,38 @@ public void refresh() {
     }
 
     /**
+     * Blocks the current thread until new changes to the Realm are available or {@link #stopWaitForChange}
+     * is called from another thread.
+     *
+     * @return {@code true} If the Realm was updated to the latest version, {@code false} if it was
+     * cancelled by calling {@link #stopWaitForChange}.
+     * @throws IllegalStateException if attempting to wait within a transaction or a Looper thread.
+     */
+    public boolean waitForChange() {
+        checkIfValid();
+        if (isInTransaction()) {
+            throw new IllegalStateException("Cannot wait for changes inside of a transaction.");
+        }
+        if (Looper.myLooper() != null) {
+            throw new IllegalStateException("Cannot wait for changes inside a Looper thread. Use RealmChangeListeners instead.");
+        }
+        boolean hasChanged = sharedGroupManager.getSharedGroup().waitForChange();
+        if (hasChanged) {
+            sharedGroupManager.advanceRead();
```

Comment:
That was recently changed. `sync_if_needed` is only called manually now (in order to support stable iterators). Also RealmResults save the last version they have seen. So you have to go through `refreshSynchronousTableViews` to keep the invariants in place.